### ISSUE TITLE
What do you think of moving to C++17

### ISSUE THREAD

**Issue by shrit** [mlpack/mlpack#2996]
Hello everyone,

Boost has a lot of nice features that helped in accelerating the development of mlpack. It has also a lot of functionalities that are developed independently and then integrated into the C++ standard library. However, recently, we have decided to remove boost from mlpack in order to reduce dependencies overlead, and to increase mlpack compilation time.
However, during the removal of Boost, several features are available there but not available in the current standard library that mlpack uses (which is C++11) such as `std::any`, or `std::string_view`. In addition, it will be too complicated to write such features, or incorporated the current implementation into the source code.

I think it is the time to start moving to C++17, allowing slowly to integrate the new features into the code base for the future release of mlpack. The main issue that might arise is the compilation and features support by the list of available compilers, according to this [list](https://en.cppreference.com/w/cpp/17),  C++17 features are included in these compilers starting from these versions:

```
gcc   >= 7
clang >= 6
MSVC  >= 19.14
```
In Linux, most of the current distributions have gcc 7, even old Ubuntu (16) which is no longer supported and dates back 5 years ago can have gcc 7 installed. I think it is the same for clang 6 which is available in Linux, and MacOS. However, I am not familiar with MSVC and I do not know which features are supported and which are not, even according to the this [list](https://en.cppreference.com/w/cpp/17) it was hard to match the version mentioned in Wikipedia and the available one.

The idea is to start moving gradually to C++17, by replacing boost part with the one available in standard library, and incorporated C++17 features slowely in the code base. The objective can be achieved (hopefully) when mlpack 4.0 will be released.

Let me know what do you think about this proposal, It is a good time to discuss this in the comments

PS: If anyone has Visual Studio, it might be a good idea to tell us in the comment, what are the C++17 features that are supported, and the compiler version that is currently being used.

**Comment by zoq** [comment]
The main question I have is what does C++17 bring to the table, like the two features you mentioned above could be replaced by:

- https://github.com/mlpack/ensmallen/blob/master/include/ensmallen_bits/utility/any.hpp
- https://github.com/bitwizeshift/string_view-standalone

without depending on C++17. I'm not against bumping up the version, but if there is a way to be C++11 complaint without sacrificing anything really I'm not sure I see the reason to go that route at least not yet.

**Comment by shrit** [comment]
There are actually a set of great features that we can use with C++17, for instance, we can have the compile-time if, which will simply mlpack code, for instance, we can use `if constexpr (is_same<A, B>::value)`, instead of having several overloads with `typename std::enable_if<is_same>....`.

Of course, if you think that we are going to lose a considerable number of users because of this shift, then we can wait. But if most users have a set of compilers that can support these features then we will stay in the past for no specific reason.

Finally, the library you have pointed out is great https://github.com/bitwizeshift/BackportCpp, we can use it directly in mlpack to resolve the issue related to boost and then move in the future to C++17. But the question is, when do you think is the time to do this shift. 

Also, I am wondering if we will have new users if we use C++17. Would some folks be interested to use a machine library that uses these features, this is still another question to think about.

**Comment by zoq** [comment]
Personally I would switch if there is a feature that isn't only sugar on top of something we already use in some form or the other that would make our life easier or currently isn't possible.

armadillo currently sticks with C++11 as well, not sure this is because of limited compiler support or something else. @conradsnicta might have some more insights.

**Comment by conradsnicta** [comment]
> armadillo currently sticks with C++11 as well, not sure this is because of limited compiler support or something else.

This is mainly due to the need to support RHEL-7 / CentOS-7, which has gcc 4.8 as the default compiler.  RHEL-7 / CentOS-7 are (still) used on clusters in academic and commercial organisations.  There is an overall (slow) transition to RHEL-8, though some places are simply waiting for RHEL-9 (due in 2022).

Purely from the perspective of Armadillo, right now I don't have a pressing need to use specific C++17 features.  C++17 does have the "[if constexpr](https://en.cppreference.com/w/cpp/language/if)" feature which may simplify a few things, but it's not critical.

The next logical step for Armadillo seems to be C++20, which has the "[concepts](https://en.wikipedia.org/wiki/Concepts_(C%2B%2B))" feature that is very useful for code involving lots of templates.  gcc 10 has almost full support for concepts, as per https://gcc.gnu.org/projects/cxx-status.html and https://en.cppreference.com/w/cpp/compiler_support. 

To my understanding, gcc 10 will be part of the next Debian release (due in Q3 2021).  Either gcc 10 or gcc 11 is going to be part of RHEL-9 and Ubuntu 22.04 LTS.

**Comment by rcurtin** [comment]
I've always been a fan of maintaining reverse compatibility, so I have tried to keep our minimum versions as low as possible---so long as it doesn't require *too* much effort from our end.  (And I think in the past, there are places where I did expend too much effort on this goal...)

I agree with @conradsnicta that it's important to support whatever the oldest-widely-used-version of RHEL, especially for institutions or universities, where individual users may not have control over their systems but may instead get accounts on university-maintained RHEL systems.

In our case there appear to be two main things we would like from C++17:

 - `std::any`
 - `std::string_view`

Primarily, the reason we would like them is to replace Boost.

At least personally, I think the route @zoq proposed is pretty reasonable: use backported/adapted versions of those two things if C++17 is unavailable.  The version of `any` provided by ensmallen is *probably* sufficient for our needs (I think it is only used in the binding support code?), and with `string_view`, it seems like maybe if we don't have C++17 support, we can use a backported version, or use regular `std::string_view` otherwise.  That can probably be controlled based on macros...

RHEL7 doesn't go EOL until 2024... so we might have the code for backport support for some years.  I guess we can see as time goes on what the continuing usage of RHEL7 is?

**Comment by conradsnicta** [comment]
> I agree with @conradsnicta that it's important to support whatever the oldest-widely-used-version of RHEL, especially for institutions or universities, where individual users may not have control over their systems but may instead get accounts on university-maintained RHEL systems.

cf. https://github.com/mlpack/mlpack/pull/2955 and the related discussion at https://github.com/mlpack/mlpack/issues/2939

TLDR: mlpack no longer supports plain RHEL7 (ie. gcc 4.8) and requires a newer compiler which can be installed on RHEL7.

**Comment by shrit** [comment]
According to the above opinions, let us initially go with the backported version [`bpstd`](https://github.com/bitwizeshift/BackportCpp).
If `bpstd` compiles well, we can use it for the next couple of years, I would say 2024.
In the meanwhile, I will start replacing the `boost::any_cast` and see if this compiles with `bpstd` version. I will include the necessary `bpstd` headers inside mlpack since no one have them on their machine and they are intended to be used in such scenario.

**Comment by rcurtin** [comment]
That sounds like a good plan. :+1:

@conradsnicta is right---mlpack will no longer compile on the standard GCC provided by RHEL7.  Somehow I forgot that in my response.  Instead users will have to install `devtoolset-8`, which provides gcc 8.2.1.  It looks like that will have everything we need from C++17 support, but, we still have to also consider Ubuntu (18.04), which has gcc 7.4.0.  Maybe that is new enough?

Anyway, with that bit about RHEL7 in mind, I'm ok with either approach... the backport approach allows our users to stick with older compilers if they have them for some reason, but I think C++17 support should be stable enough for most users---and for those on RHEL7, they already need `devtoolset-8` anyway.

**Comment by conradsnicta** [comment]
According to gcc documentation, almost all C++17 features are present in gcc 7:
* https://gcc.gnu.org/projects/cxx-status.html#cxx17
* https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017

The main omission seems to be [filesystem library](https://en.cppreference.com/w/cpp/filesystem).

**Comment by rcurtin** [comment]
We talked about it in the video chat today; @zoq pointed out that if Armadillo only requires a C++11 compiler, we should probably try and keep the same requirement for mlpack.  I can agree with this (but my opinion is not particularly strong).

So I think the idea @shrit is going with, where we are backporting support when C++17 isn't available, instead of just requiring C++17, is a good approach.
